File: src/audio_steg_app/lib/app/app_config.dart
import 'dart:convert';

import 'package:flutter/services.dart';

/// Deploy-time configuration from repo-root [appsettings.json] (not user preferences).
class AppConfig {
  final bool showEmbedLoadFileButton;
  final bool showEmbedRecoveryDialog;

  const AppConfig({
    required this.showEmbedLoadFileButton,
    required this.showEmbedRecoveryDialog,
  });

  static const AppConfig defaults = AppConfig(
    showEmbedLoadFileButton: false,
    showEmbedRecoveryDialog: true,
  );

  static Future<AppConfig> load() async {
    try {
      final raw = await rootBundle.loadString('../../appsettings.json');
      final map = jsonDecode(raw) as Map<String, dynamic>;
      return AppConfig(
        showEmbedLoadFileButton:
            map['ShowEmbedLoadFileButton'] as bool? ??
            defaults.showEmbedLoadFileButton,
        showEmbedRecoveryDialog:
            map['ShowEmbedRecoveryDialog'] as bool? ??
            defaults.showEmbedRecoveryDialog,
      );
    } catch (_) {
      return defaults;
    }
  }
}

File: src/audio_stegano_desktop/src/AudioStegano.Desktop/AppConfig.cs
using System.IO;
using System.Text.Json;

namespace AudioStegano.Desktop;

/// <summary>
/// Deploy-time flags from <c>appsettings.json</c> beside the executable (not user settings UI).
/// </summary>
public sealed class AppConfig
{
    /// <summary>When false, hides the upload-file button on the embed (steganography) tab.</summary>
    public bool ShowEmbedLoadFileButton { get; set; } = false;

    /// <summary>When true, shows recovery message-length dialog after embed.</summary>
    public bool ShowEmbedRecoveryDialog { get; set; } = true;

    /// <summary>Fixed msg_len when default message bit limit is enabled (2^18 = 262144).</summary>
    public int DefaultFixedMessageBitLength { get; set; } = 262144;

    /// <summary>Default logistic map r for new sessions / reset.</summary>
    public double LogisticR { get; set; } = 3.99;

    /// <summary>Default logistic map x0 for new sessions / reset.</summary>
    public double LogisticX0 { get; set; } = 0.45;

    /// <summary><c>xor_only</c> or <c>ae_xor</c> — <c>train/embed_message.m</c> embed_mode.</summary>
    public string DefaultStegoEmbedMode { get; set; } = "xor_only";

    public static AppConfig Current { get; private set; } = new();

    /// <summary>Embed tab: on Windows desktop, file load stays available even when deploy flag is false (mobile-oriented default).</summary>
    public static bool ShowEmbedLoadFileForUi =>
        Current.ShowEmbedLoadFileButton || OperatingSystem.IsWindows();

    private static string AppSettingsFilePath =>
        Path.Combine(AppContext.BaseDirectory, "appsettings.json");

    public static void Load()
    {
        Current = new AppConfig();
        if (!File.Exists(AppSettingsFilePath)) return;
        try
        {
            var json = File.ReadAllText(AppSettingsFilePath);
            var loaded = JsonSerializer.Deserialize<AppConfig>(json);
            if (loaded is not null)
                Current = loaded;
        }
        catch
        {
            // Keep defaults when the file is invalid.
        }
    }
}
